### DISCUSSION TITLE
Text Recognition Fails on Rotated Text (Architectural Drawings) – Detection Box Corner Order Seems Wrong

### DISCUSSION THREAD

**Discussion by bananaback** [PaddlePaddle/PaddleOCR#17325] [Q&A]
Hi developers,
I’m working on an OCR system for architectural drawings, where text appears at many different angles (not only horizontal).

Horizontal text works fine, but rotated text often becomes flipped or mis-read.
Here is an example of how the rotated labels look in my dataset:
(see image)
![image (1)](https://github.com/user-attachments/assets/25ce94df-ee5e-4fc4-a86d-a9f28f9a8bdd)

And here is how the recognition goes wrong when the text rotates further:
<img width="762" height="417" alt="image (3)" src="https://github.com/user-attachments/assets/2b671703-edc8-4008-a0b3-2df0bed6fdb9" />

### 1. Finetuning the recognition model

I first tried finetuning the recognition model, but it did not help.
So I moved on to investigating the detection stage.

### 2. Investigating the detection model using PPOCRLabel

During annotation in PPOCRLabel, I noticed something important:

When PPOCRLabel auto-draws the box (top-left → bottom-right),
→ Recognition is wrong for rotated text.
<img width="821" height="389" alt="image (4)" src="https://github.com/user-attachments/assets/22c7f53b-7d46-4f5d-96b3-1e83115fbb20" />

When I manually draw the box carefully from bottom-left → top-right,
then click Re-Recognition,
→ The recognition becomes correct.
<img width="697" height="316" alt="image (5)" src="https://github.com/user-attachments/assets/373119da-ed62-4ab1-89a0-28643cdffd77" />

This made me think the ordering of the 4 points affects the angle/direction, so I labeled my rotated dataset very carefully using this bottom-to-top direction.

I labeled ~500 rotated samples and finetuned the detection model.

### 3. Testing the finetuned detection model

After finetuning, the real prediction results are still not good — rotated text is still mis-recognized.

So I inspected the output boxes from predict(), and this is what I found:

Every predicted box always follows the same point order

top-left

top-right

bottom-right

bottom-left

This happens for all rotations — vertical, diagonal, upside down, etc.
It seems the detector normalizes every box back into TL–TR–BR–BL and does not preserve any rotation information.

Because of this, the recognition model receives no real “angle” signal.
<img width="2048" height="2048" alt="Gemini_Generated_Image_igvq8gigvq8gigvq" src="https://github.com/user-attachments/assets/45ff221a-bf81-40a8-9541-9b662a2d0c6c" />

### 1. How can I correctly finetune the detection model for rotated text?

In my dataset, text can rotate freely clockwise.
But when it rotates beyond 270° clockwise, it effectively becomes a 90° counter-clockwise rotation (as shown in the first attached image).

During labeling, for vertical or highly rotated text, I manually draw the box from bottom-left → top-right, then click Re-Recognition — and the recognition becomes correct.
This is why I believed the point order helps the model learn direction.

I labeled ~500 such rotated samples carefully, finetuned the detection model, but the final predictions are still angle-neutral (always TL → TR → BR → BL).
So the recognition model never receives orientation information.

I’m not sure whether:

1. my finetuning approach is wrong,
2. my dataset is too small,
3. or the detection model simply cannot learn angle information at all.

### 2. Which module in PaddleOCR is actually responsible for orientation?

Because the detection outputs always normalized corner order, I am unsure where rotation is supposed to be handled:

- Is angle predicted by another module?
- Is there an angle classifier that needs to be trained?
- Or is the recognition model expected to infer rotation directly from the cropped patches?
- Am I misunderstanding how the PaddleOCR pipeline handles orientation?

I’ve searched documentation and online discussions for days, but I still can’t find a clear explanation of which component is responsible for rotation or what the correct training process is for multi-angle text.

Right now I don’t know whether:

1. I trained the wrong module,
2. I labeled incorrectly,
3. or I misunderstood the intended pipeline.

**Comment by liuhongen1234567** [comment]
Hello, I tested the demo image and found that the recognition results are not particularly problematic. Below is my inference code:
```python
from paddleocr import PaddleOCR

ocr = PaddleOCR(
    use_doc_orientation_classify=False, # 通过 use_doc_orientation_classify 参数指定不使用文档方向分类模型
    use_doc_unwarping=False, # 通过 use_doc_unwarping 参数指定不使用文本图像矫正模型
    use_textline_orientation=True, # 通过 use_textline_orientation 参数指定不使用文本行方向分类模型
)
result = ocr.predict("/paddle/ppocr_vl_model/issue_test/dd893599980caf804c1d85ebb7ce5581.png")
for res in result:
    res.print()
    res.save_to_img("output")
    res.save_to_json("output")
```
<img width="1196" height="416" alt="dd893599980caf804c1d85ebb7ce5581_ocr_res_img" src="https://github.com/user-attachments/assets/48c992c5-f6d0-4c3e-b819-15c6082ba6aa" />

**Reply by bananaback** [reply]
Hi liuhongen,

Thank you very much for taking the time to test the demo image and for sharing your inference code — I really appreciate your reply.

Sorry for the late response. Our team has recently been busy preparing a demo, so things have been a bit hectic.

Now I would like to explain more clearly the issue I’m currently facing. After spending several hours reading the PaddleOCR documentation and going through the source code, I still have some questions and points that I hope you could help clarify.

**Reply by bananaback** [reply]
As I understand it, PaddleOCR first detects text boxes, then uses a projection operation to convert the polygon into a rectangular bounding box. After that, it automatically rotates the image 90 degrees counterclockwise if the height is greater than 1.5 times the width.

Next, the cropped text image is passed into the text line orientation model to decide whether it should be rotated another 180 degrees (if the user enables text line orientation).

Finally, the recognition model always receives text images in a horizontal orientation.
<img width="1334" height="736" alt="Screenshot 2025-12-18 144030" src="https://github.com/user-attachments/assets/3b18c028-e152-4135-beb9-6a722de80135" />

**Reply by bananaback** [reply]
My first attempt was to fine-tune the detection model so that it can recognize short, standalone characters, such as (E, F, H, ...), what I did was fine-tune the model by labeling the bounding boxes wider on both sides instead of tightly wrapping the characters.

The purpose was to make the detected text boxes always appear horizontal, so they would not be rotated by PaddleOCR’s default behavior.
<img width="1367" height="826" alt="Screenshot 2025-12-18 144424" src="https://github.com/user-attachments/assets/c9fb5622-2696-4da9-b765-0a4144fc2d41" />

**Reply by bananaback** [reply]
And there is another case: for text that is originally in a vertical orientation, it gets rotated 90 degrees, which causes the recognition model to become confused.
<img width="1259" height="867" alt="Screenshot 2025-12-18 144827" src="https://github.com/user-attachments/assets/298ad536-a74b-4be0-8643-cdaf77fcea80" />

**Reply by bananaback** [reply]
<img width="1044" height="627" alt="Screenshot 2025-12-18 144902" src="https://github.com/user-attachments/assets/9dedc8c6-0616-484a-8e2f-1c12c4abaed0" />

**Reply by bananaback** [reply]
In theory, I should enable use_text_line_orientation,
and fine-tune the text line orientation model with a dataset that matches my specific use case.
<img width="1227" height="466" alt="Screenshot 2025-12-18 144944" src="https://github.com/user-attachments/assets/51b41f52-8e73-42b3-9969-1f92e554d2c3" />

**Reply by bananaback** [reply]
However, due to limited time, I used a workaround: I disabled text line orientation and modified the library code to rotate the text 90 degrees clockwise instead.

This allowed me to avoid spending a lot of time creating a dataset and fine-tuning the model.
<img width="1107" height="663" alt="Screenshot 2025-12-18 150124" src="https://github.com/user-attachments/assets/62ed709e-b8ca-4012-bddf-ad5d509fc486" />

<img width="1279" height="504" alt="Screenshot 2025-12-18 145259" src="https://github.com/user-attachments/assets/0ba44317-fbc4-40c6-bfc7-ca2ffdd1eb40" />

<img width="1191" height="554" alt="Screenshot 2025-12-18 145314" src="https://github.com/user-attachments/assets/de2723a2-0270-40ab-86f6-15358e9cfc72" />

**Reply by bananaback** [reply]
And the model works well in most cases, but there are still many difficult edge cases.

We are about to move into a more serious development phase, and I’m concerned that in the long run I may encounter cases where the text is vertically aligned downward. In such situations, this workaround would no longer work, and I’m also unsure whether my current solution is truly reliable.

I’m therefore considering a different approach: instead of customizing PaddleOCR’s source code, I would enable use_text_line_orientation and fine-tune the orientation model according to the official guidelines. In this case, I have a few questions, such as how large the dataset needs to be to reliably distinguish between 0° and 180° orientations, and what an effective fine-tuning strategy would look like.

Thank you to the developers, and I look forward to your response.

**Comment by MelloCaique** [comment]
Any update on this? I'm having the exact same problem, but with the added complication that my text can be rotated at any angle.
When using PPOCRLabel, if I draw the RectBox with the right text orientation and not only TL–TR–BR–BL as mentioned by @bananaback, the the text is extracted correctly. The question is:

- How can I properly fine-tune the rotation model?
